### PR TITLE
Automatically determine image/tensor color mapping & need for srgb decoding

### LINKED ISSUE #2274: Image viewer normalizes (?) the colors and popup shows unnormalized RGB
**Describe the bug**
```python
import rerun as rr
import numpy as np
import matplotlib.pyplot as plt

rr.init("image color", spawn=True)
img = np.random.random((300, 300, 3)) * 0.2
rr.log_image("image", img)
print(img.min(), img.max())  # to check they are in range [0,1]

plt.imshow(img)
plt.show()
```
gives
image_color

Hovering over the image in rerun shows low RGB values, but bright colors.

**Expected behavior**
I'd expect it to look the same / very similar with the shown RGB values being the same as those in the raw data.

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04

**Rerun version**
```
rerun_py 0.7.0-alpha.0+ [rustc 1.69.0 (84c898d65 2023-04-16), LLVM 15.0.7] x86_64-unknown-linux-gnu main b263b9f, built 2023-05-29T18:38:22Z
```
Also happens in `0.6`

### PR DESCRIPTION
**Pull request by Wumpf**
### What

Fixes #2274 
* #2274 

Will need to be further worked on via:
* https://github.com/rerun-io/rerun/issues/2341

Introduces 3 new things actually:
* finite range, determined if the range of a tensor was in fact not finite
* heuristic for color mapping range
* heuristic for srgb
   * ⚠️ does this break things?

Range & sRGB:

This
```
import rerun as rr
import numpy as np
import matplotlib.pyplot as plt

rr.init("image color", spawn=True)
img = np.random.random((300, 300, 3)) * 0.2
rr.log_image("image", img)
print(img.min(), img.max())  # to check they are in range [0,1]

plt.imshow(img)
plt.show()
```
Gives now the expected image:
image

When setting a pixel to inf, we show the "finite range" as well, and behave correctly:
image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)

PR Build Summary: https://build.rerun.io/pr/2342

Docs preview: https://rerun.io/preview/a1011c1/docs
Examples preview: https://rerun.io/preview/a1011c1/examples

### CODE REVIEW

**Comment by Wumpf** [comment]
`python ./examples/python/api_demo/main.py --demo image_tensors`
before:
image

after:
image

looks the same which is encouraging

**Review by emilk — CHANGES_REQUESTED** [review]
I think the heuristics are good (except for it being RGB(A) only), but the I would expect it to have some effect on our tests. In particular, I would expect it would make all float and unsigned types in `./examples/python/api_demo/main.py --demo image_tensors` have the correct gamma.

File: crates/re_renderer/shader/rectangle_fs.wgsl [hunk 1]
```diff
@@ -110,6 +110,8 @@ fn fs_main(in: VertexOut) -> @location(0) Vec4 {
         return ERROR_RGBA; // unknown color mapper
     }
 
+    texture_color = select(texture_color, Vec4(1.0), texture_color > 1.0);
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_renderer/shader/rectangle_fs.wgsl:113]: ehm, full disclosure, I accidentally left my debugging journey in here 😳. Needs removing

File: crates/re_viewer_context/src/gpu_bridge/mod.rs [hunk 2]
```diff
@@ -25,28 +25,59 @@ pub enum RangeError {
...
     let max = max as f32;
 
-    if !min.is_finite() || !max.is_finite() {
-        Err(RangeError::NonfiniteRange)
+    // Apply heuristic for ranges that are typically expected depending on the data type and the finite (!) range.
+    // (we ignore NaN/Inf values heres, since they are usually there by accident!)
+    if data_type.is_float() && min >= 0.0 && max <= 1.0 {
+        // Float values that are all between 0 and 1, assume that this is the range.
+        Ok([0.0, 1.0])
+    } else if min >= 0.0 && max <= 255.0 {
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_viewer_context/src/gpu_bridge/mod.rs:45]: it's not numbers aren't on a line, it's just that constants are always on the right side in my head for no particular reason and if I don't think about it that's stronger than line-ness 🤷

File: crates/re_viewer_context/src/tensor/tensor_stats.rs [hunk 3]
```diff
@@ -61,6 +64,41 @@ impl TensorStats {
...
+            re_tracing::profile_function!();
+            let (min, max) = tensor.fold((f16::MAX, f16::MIN), |(min, max), &value| {
+                if value.is_finite() {
+                    (min.min(value), max.max(value))
+                } else {
+                    (min, max)
+                }
+            });
+            (min.to_f64(), max.to_f64())
+        }
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_viewer_context/src/tensor/tensor_stats.rs:100]: makes sense, I like this

File: crates/re_viewer_context/src/gpu_bridge/mod.rs [hunk 4]
```diff
@@ -25,28 +25,59 @@ pub enum RangeError {
...
     }
 }
 
+/// Return whether a tensor should be assumed to be encoded in sRGB color space ("gamma space", no EOTF applied).
+pub fn tensor_decode_srgb_heuristic(
+    tensor_stats: &TensorStats,
+    data_type: re_components::TensorDataType,
+    channels: u32,
+) -> Result<bool, RangeError> {
+    if channels >= 3 {
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_viewer_context/src/gpu_bridge/mod.rs:64]: let's also name it `tensor_decode_srgb_eotf_heuristic` to be explicit then!

**Comment by Wumpf** [comment]
Much much better now! Needed quite a few more changes, but I'm happy with those now
image

File: crates/re_renderer/shader/rectangle_fs.wgsl [hunk 5]
```diff
@@ -14,78 +14,81 @@ fn tex_filter(pixel_coord: Vec2) -> u32 {
...
+    return (sampled_value - range.x) / (range.y - range.x);
+}
 
-    // Convert to linear space:
-    if rect_info.decode_srgb != 0u {
+fn decode_color(sampled_value: Vec4) -> Vec4 {
+    // Normalize the value first, otherwise premultiplying alpha and linear space conversion won't make sense.
+    var rgba = normalize_range(sampled_value);
+
+    // Convert to linear space. Skip if values are not in the 0..1 range (might be inf).
```
Comment (emilk) [inline crates/re_renderer/shader/rectangle_fs.wgsl:26]: We could use `ERROR_RGBA` for such values!

**Review by emilk — APPROVED** [review]
Looks good, but please make sure you test it with depth maps and tensor view before merging!

**Comment by Wumpf** [comment]
oh. yes. haven't done that yet!

File: crates/re_renderer/shader/rectangle_fs.wgsl (hunk 5 above)
Comment (Wumpf) [reply crates/re_renderer/shader/rectangle_fs.wgsl:26]: good idea, now that the normalization happens always before, we can properly error out

**Comment by Wumpf** [comment]
looked all good, didn't find any regressions!

**Pull request merged**